File: springboot/distribusi.stok.barang/src/main/java/be/distribusi/stok/barang/service/order/OrderService.java
package be.distribusi.stok.barang.service.order;

import be.distribusi.stok.barang.dto.order.ReqOrderDTO;
import be.distribusi.stok.barang.dto.order.ResOrderDTO;
import be.distribusi.stok.barang.exception.AppException;
import be.distribusi.stok.barang.model.EntityBarang;
import be.distribusi.stok.barang.repository.BarangRepository;
import be.distribusi.stok.barang.repository.ErrorRepository;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

@Service
@Slf4j
@RequiredArgsConstructor
public class OrderService {

    private final ErrorRepository errorConfig;
    private final BarangRepository barangRepository;
    //Order Barang Service
    public ResponseEntity<?> orderBarangService(@Valid ReqOrderDTO requestBody){
        // Buat objek LocalDateTime
        LocalDateTime now = LocalDateTime.now();
        // Format LocalDateTime menjadi String
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
        String loggerId = now.format(formatter);

        String kodeBarang = requestBody.getKodeBarang().toUpperCase();
        String namaBarang = requestBody.getNamaBarang().toUpperCase();
        Integer banyakPesanan = Integer.parseInt(requestBody.getBanyakPesanan());
        //Get data Barang
        List<EntityBarang> barangList = barangRepository.findBarangbyCodeName(kodeBarang, namaBarang);
        //Cek Barang is Exist
        if (barangList.isEmpty()){
            log.error("{} - Barang tidak ditemukan", loggerId);
            String statusCode = errorConfig.getErrorCode("item_nfound");
            String errorMessage = errorConfig.getErrorMessage("item_nfound");
            throw new AppException(statusCode, errorMessage, HttpStatus.BAD_REQUEST, loggerId);
        }
        EntityBarang barang = barangList.get(0);
        //Update srok Keluar dan sisa stok
        Integer stokKeluarTemp = (barang.getStokKeluar() == null || barang.getStokKeluar().equals("")) ? 0 : barang.getStokKeluar();
        int sisaStok = (barang.getSisaStok() - banyakPesanan);
        int stokKeluar = stokKeluarTemp + banyakPesanan;
        barangRepository.updateOrderBarang(sisaStok, stokKeluar, kodeBarang, namaBarang);
        // Hitung Harga Pesanan
        BigDecimal banyakPesananDecimal = BigDecimal.valueOf(banyakPesanan);
        BigDecimal totalHarga = barang.getHargaJual().multiply(banyakPesananDecimal);
//        Integer totalHarga = (banyakPesanan * barang.getHargaJual());

        // Mapping request ke responseDTO
        ResOrderDTO.orderDataDTO orderDataDTO = new ResOrderDTO.orderDataDTO();
        // Set data ke responseDTO
        orderDataDTO.setLoggerId(loggerId);
        orderDataDTO.setKodeBarang(kodeBarang);
        orderDataDTO.setNamaBarang(namaBarang);
        orderDataDTO.setHargaJual(String.valueOf(barang.getHargaJual()));
        orderDataDTO.setTotalPesanan(String.valueOf(banyakPesanan));
        orderDataDTO.setTotalHarga(totalHarga.toString());

        ResOrderDTO responseDTO = new ResOrderDTO();
        responseDTO.setData(orderDataDTO);
        responseDTO.setStatusCode("200");
        responseDTO.setStatusMsg("SUCCESS");
        responseDTO.setErrorMessage("");

        log.info("{} - End-to-End processing complete.", loggerId);

        return ResponseEntity.status(HttpStatus.OK).body(responseDTO);
    }
}

File: python/distribusi-stok-barang/services/orderService.py
from datetime import datetime
from decimal import Decimal
from models.databaseModels import FindBarangbyCodeName, UpdateOrderBarang
from helpers.config import Config as config
from exception.appException import AppException
from loguru import logger as log
import json

def OrderDistribusiBarang(req, db):
    result = {
        "data":{},
        "status_code": "",
        "status_msg": "",
        "error_msg": "",
    }
    loggerId = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    log.info(f"{loggerId} - Start to processing select service. Request: \n{json.dumps(dict(req))}")

    kodeBarang = req.kode_barang.upper()
    namaBarang = req.nama_barang.upper()
    banyakPesanan = int(req.banyak_pesanan)

    listBarang = FindBarangbyCodeName(db, kodeBarang, namaBarang)
    
    if not listBarang:
        log.error(f"{loggerId} - Barang tidak ditemukan")
        raise AppException(loggerId, config.param_error['item_nfound'][0], config.param_failed, config.param_error['item_nfound'][1])
    
    barang = listBarang[0]
    
    stokKeluarTemp = 0 if barang['stok_keluar'] in [None, ''] else barang['stok_keluar'] 

    sisaStok = (barang['sisa_stok'] - banyakPesanan)
    stokKeluar = (stokKeluarTemp + banyakPesanan)

    UpdateOrderBarang(
        db,
        kodeBarang,
        namaBarang,
        sisaStok,
        stokKeluar
    )
    banyakPesananDecimal = Decimal(banyakPesanan)
    totalHarga = (barang['harga_jual'] * banyakPesananDecimal)
    
    data = {
        "logger_id" : loggerId,
        "kode_barang": kodeBarang,
        "nama_barang": namaBarang,
        "harga_jual": str(barang['harga_jual']),
        "total_pesanan": str(banyakPesanan),
        "total_harga": str(totalHarga)
    }
    result["data"] = data
    result["status_code"] = "200"
    result["status_msg"] = config.param_success
    result["error_msg"] = ""
    log.info(f"{loggerId} - End-to-End processing complete. Result: \n{json.dumps(dict(result))}")
    return result

    
